akka-persistence/src/main/scala/akka/persistence/journal/EventAdapter.scala

```diff
@@ -0,0 +1,91 @@
+ *
+ * Typical use cases include (but are not limited to):
+ * <ul>
+ *   <li>adding metadata, a.k.a. "tagging" - by wrapping objects into tagged counterparts</li>
+ *   <li>manually converting to the Journals storage format, such as JSON, BSON or any specialised binary format</li>
+ *   <li>adapting incoming events in any way before persisting them by the journal</li>
+ * </ul>
+ */
+abstract class EventAdapter {
+  //#event-adapter-api
+  /**
+   * Return the manifest (type hint) that will be provided in the `fromJournal` method.
+   * Use `""` if manifest is not needed.
+   */
+  def manifest(event: Any): String
+
+  /**
+   * Convert domain event to journal event type.
+   *
+   * Some journal may require a specific type to be returned to them,
+   * for example if a primary key has to be associated with each event then a journal
+   * may require adapters to return `com.example.myjournal.EventWithPrimaryKey(event, key)`.
+   *
+   * The `toJournal` adaptation must be an 1-to-1 transformation.
+   * It is now allowed to drop incoming events during the `toJournal` adaptation.
+   *
+   * @param event the application-side domain event to be adapted to the journal model
+   * @return the adapted event object, possibly the same object if no adaptation was performed
+   */
+  def toJournal(event: Any): Any
+
+  /**
+   * Convert a event from its journal model to the applications domain model.
+   *
+   * One event may be adapter into multiple (or none) events which should be delivered to the [[akka.persistence.PersistentActor]].
+   * Use the specialised [[EventSeq.single]] method to emit exactly one event, or [[EventSeq.empty]] in case the adapter
+   * is not handling this event. Multiple [[EventAdapter]] instances are applied in order as defined in configuration
+   * and their emitted event seqs are concatenated and delivered in order to the PersistentActor.
+   *
+   * @param event event to be adapted before delivering to the PersistentActor
+   * @param manifest optionally provided manifest (type hint) in case the Adapter has stored one for this event, `""` if none
+   * @return sequence containing the adapted events (possibly zero) which will be delivered to the PersistentActor
+   */
+  def fromJournal(event: Any, manifest: String): EventSeq
+
+  //#event-adapter-api
+}
+
+sealed abstract class EventSeq {
+  def events: immutable.Seq[Any]
+}
+object EventSeq {
+  /** Java API */
+  final def empty: EventSeq = EmptyEventSeq
+
+  /** Java API */
+  final def single(event: Any): EventSeq = new SingleEventSeq(event)
+
+  /** Java API */
+  @varargs final def create(events: Any*): EventSeq = EventsSeq(events.toList)
+  final def apply(events: Any*): EventSeq = EventsSeq(events.toList)
+}
+final case class SingleEventSeq(event: Any) extends EventSeq { // TODO try to make it a value class, would save allocations
+  override def events: immutable.Seq[Any] = List(event)
```

Comment:
Thanks. 
I was trying to make this one a value class, however things get complicated then sadly...
(`EventSeq` must be an interface and must `extend Any` (universal trait), which destroys the Java API forwarders etc). I hope to try this again, we could save on some allocations here.